### PR TITLE
[#17184]: Advertise X-OAUTH as a sasl mechanism only if mod_auth_token module is loaded.

### PR DESCRIPTION
**Pull request by epretha**
This PR addresses #1312 

Proposed changes include:
* The cyrsasl.erl module has been updated to filter out X-OAUTH sasl mechanism if the mod_auth_token.erl module is not loaded.
* The xouath_SUITE.erl test suite has been updated with a new test case.

### CODE REVIEW

File: apps/ejabberd/src/cyrsasl.erl [hunk 1]
```diff
@@ -126,7 +126,8 @@ listmech(Host) ->
                                  []
                          end,
                          ['$1']}]),
-    filter_anonymous(Host, Mechs).
+    filter_mechanisms(Host, Mechs, [{<<"ANONYMOUS">>, fun(H) -> ejabberd_auth_anonymous:is_sasl_anonymous_enabled(H) end},
```
Comment (gadgetci) [inline apps/ejabberd/src/cyrsasl.erl]: According to **Elvis**:
> Line 129 is too long:     filter_mechanisms(Host, Mechs, [{<<"ANONYMOUS">>, fun(H) -> ejabberd_auth_anonymous:is_sasl_anonymous_enabled(H) end},.

File: apps/ejabberd/src/cyrsasl.erl [hunk 2]
```diff
@@ -126,7 +126,8 @@ listmech(Host) ->
...
+                                    {<<"X-OAUTH">>, fun(H) -> gen_mod:is_loaded(H, mod_auth_token) end}]).
```
Comment (gadgetci) [inline apps/ejabberd/src/cyrsasl.erl]: According to **Elvis**:
> Line 130 is too long:                                     {<<"X-OAUTH">>, fun(H) -> gen_mod:is_loaded(H, mod_auth_token) end}])..

File: test.disabled/ejabberd_tests/tests/oauth_SUITE.erl [hunk 3]
```diff
@@ -313,6 +315,26 @@ provision_token_login(Config) ->
...
+    ConnSteps = [start_stream,
+                 stream_features,
+                 maybe_use_ssl,
+                 maybe_use_compression],
+    {ok, _, Features} = escalus_connection:start(AliceSpec, ConnSteps),
+    true = lists:member(<<"X-OAUTH">>, proplists:get_value(sasl_mechanisms,
+                                                           Features, [])),
+    dynamic_modules:stop(Host, Module),
+    {ok, _, NewFeatures} = escalus_connection:start(AliceSpec, ConnSteps),
+    false = lists:member(<<"X-OAUTH">>, proplists:get_value(sasl_mechanisms,
```
Comment (fenek) [inline test.disabled/ejabberd_tests/tests/oauth_SUITE.erl]: It should be a separate test case, so there would be distinct two: positive and negative.

File: test.disabled/ejabberd_tests/tests/oauth_SUITE.erl [hunk 4]
```diff
@@ -313,6 +315,26 @@ provision_token_login(Config) ->
...
+    Module = mod_auth_token,
+    AliceSpec = escalus_users:get_userspec(Config, alice),
+    ConnSteps = [start_stream,
+                 stream_features,
+                 maybe_use_ssl,
+                 maybe_use_compression],
+    {ok, _, Features} = escalus_connection:start(AliceSpec, ConnSteps),
+    true = lists:member(<<"X-OAUTH">>, proplists:get_value(sasl_mechanisms,
+                                                           Features, [])),
+    dynamic_modules:stop(Host, Module),
```
Comment (fenek) [inline test.disabled/ejabberd_tests/tests/oauth_SUITE.erl]: This should be performed in `init_per_testcase`.

File: test.disabled/ejabberd_tests/tests/oauth_SUITE.erl [hunk 5]
```diff
@@ -313,6 +315,26 @@ provision_token_login(Config) ->
...
+    {ok, _, NewFeatures} = escalus_connection:start(AliceSpec, ConnSteps),
+    false = lists:member(<<"X-OAUTH">>, proplists:get_value(sasl_mechanisms,
+                                                            NewFeatures, [])),
+    AuthOpts = [{ {validity_period, access}, {60, minutes} },
```
Comment (fenek) [inline test.disabled/ejabberd_tests/tests/oauth_SUITE.erl]: Options should be extracted via RPC in `init_per_testcase` (what if default config changes but options here won't get an update?) and appended to test `Config`.

File: test.disabled/ejabberd_tests/tests/oauth_SUITE.erl [hunk 6]
```diff
@@ -313,6 +315,26 @@ provision_token_login(Config) ->
...
+                { {validity_period, refresh}, {1, days} }],
+    dynamic_modules:start(Host, Module, AuthOpts).
```
Comment (fenek) [inline test.disabled/ejabberd_tests/tests/oauth_SUITE.erl]: This should be performed in `end_per_testcase` with options found in `Config`.

**Comment by fenek** [comment]
Example of options backup in tests:
https://github.com/esl/MongooseIM/blob/master/test.disabled/ejabberd_tests/tests/pubsub_SUITE.erl#L214-L217
https://github.com/esl/MongooseIM/blob/master/test.disabled/ejabberd_tests/tests/pubsub_SUITE.erl#L221-L224

**Pull request merged**